File: RemoveDuplicatesFromSortedArray.java
/**
 * Created by yhf on 15/4/17.
 */

/*
Given a sorted array, remove the duplicates in place such that each element appear only once and return the new length.

Do not allocate extra space for another array, you must do this in place with constant memory.

For example,
Given input array A = [1,1,2],

Your function should return length = 2, and A is now [1,2].
*/

public class RemoveDuplicatesFromSortedArray {
    public static int removeDuplicates(int[] A) {
        int len = A.length;
        if (len <= 1) {
            return len;
        }
        int i = 0, j = 1;
        while (j < len) {
            if (A[j] == A[i]) {
                j++;
            } else {
                A[++i] = A[j++];
            }
        }
        return i + 1;
    }

    public static void main(String[] args) {
        int[] A = {1, 1};
        System.out.println(removeDuplicates(A));
    }
}

File: Leetcode-Python/RemoveDuplicatesFromSortedArray.py
"""
Given a sorted array, remove the duplicates in place such that each element appear only once and return the new length.

Do not allocate extra space for another array, you must do this in place with constant memory.

For example,
Given input array nums = [1,1,2],

Your function should return length = 2, with the first two elements of nums being 1 and 2 respectively. It doesn't
matter what you leave beyond the new length.
"""


class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def removeDuplicates(self, nums):
        if len(nums) <= 1:
            return len(nums)
        i = 0
        j = 1
        while j < len(nums):
            if nums[i] != nums[j]:
                i += 1
                nums[i] = nums[j]
            j += 1
        return i + 1


if __name__ == '__main__':
    print Solution().removeDuplicates([])
